### DISCUSSION TITLE
A brief explanation of how to implement this extension

### DISCUSSION THREAD

**Discussion by bamboowonder** [Hubelia/directus-postgresql-json-views#2] [General]
Hi there,

I'm not clear on how this extension works, but I'd very much like to prototype with it as it matches a need we have.

Essentially we have a tonne of relational 3NF and json data that we form into views.  We use material views for the json data because it is otherwise too slow.

Can you give a brief pointer on how to implement this extension to read in a view as a collection in Directus?

Thanks a bunch

**Comment by RegisWeduLabs** [comment]
Of course - Basically: 
-Clone the repository in a subfolder in the root of a directus project (I use customExtensions)
- install npm packages from the directory where you cloned the repo.
- run yarn/npm build (there is a script in package.json that will build the extension and copy the index file it in the extensions/endpoints folder of your directus project - but it needs to be in a subfolder tough.

At this point, if you start directus, you should be able to use it.  The api path is yourdirectushostname/pg-json-views - where you can use the different functions.

As an example, if you have a collection named "inventory", to create the view you would do this:

/pg-json-views/create/inventory for non materialized views OR
/pg-json-views/create/inventory?materialized=true for materialized views.

For materialized views, I suggest you create a webhook in Directus that points to http://yourdirectus:PORT/pg-json-views/refresh/inventory whenever you save something, that way, it's easier than having to call the API every time you want to update the view.

Then, simply use the following calls to get your JSON data:

/pg-json-views/get/inventory/all to get all the data - or /pg-json-views/get/pages/findone to get one document randomly.

As explained, queries aren't implemented yet - but I am planing to do this in the next few days.  Please don't hesitate to reach out!

I'm also planning on releasing a package to make it simpler.

Thanks

**Reply by RegisWeduLabs** [reply]
And to be clear, works with Postgresql - as there are functions on the database side.  Haven't tested with any other type of dbs.

**Comment by bamboowonder** [comment]
ah ok.  so this is implemented for the api, not the app?   I'm looking for views to be integrated into the end user app.

**Reply by RegisWeduLabs** [reply]
It is in the pipeline - then again, we still need to figure out how we want to do this.  That said, it's still usable - but for the time being, manually.  It's still quite simple to use tough...

Do you have a suggestion on how you would like it to be implemented on the front end?  I don't want to over complicate things either, the extension uses the Directus data structure.

**Reply by bamboowonder** [reply]
essentially fooling directus into thinking a view is a read-only table would be perfect.   In the app ui the only obvious change would be that CUD isn't possible.  Just Read.

**Reply by RegisWeduLabs** [reply]
This is exactly what this extension is doing - it takes an existing table and creates the view based on the current structure.  While there is no UI integration yet, I still suggest you try it - it's not really difficult and even if there is a UI, in the background - it would still call the endpoint the same way as if you do it manually.
